cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1690,48 +1707,32 @@ private void delete(RunRecordDetail record) throws IOException {
     getRunRecordsTable().delete(key);
   }
 
-  /**
-   * @return run records for unfinished run ignoring version
-   */
   private Map<ProgramRunId, RunRecordDetail> getUnfinishedRuns(Set<ProgramRunId> programRunIds) throws IOException {
     List<List<Field<?>>> allKeys = new ArrayList<>();
     for (ProgramRunId programRunId : programRunIds) {
       allKeys.add(getProgramRunInvertedTimeKey(TYPE_RUN_RECORD_ACTIVE, programRunId,
-                                               RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS),
-                                               false));
+                                               RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS)));
     }
-    
-    return getRunsByKeys(allKeys);
+
+    return getRunsByFullKeys(allKeys);
   }
 
   private Map<ProgramRunId, RunRecordDetail> getCompletedRuns(Set<ProgramRunId> programRunIds) throws IOException {
     List<List<Field<?>>> allKeys = new ArrayList<>();
     for (ProgramRunId programRunId : programRunIds) {
-      // Get all keys without version
-      List<Field<?>> keysWithoutVersion = getRunRecordProgramRefPrefix(TYPE_RUN_RECORD_COMPLETED,
-                                                                       programRunId.getParent().getProgramReference());
-      // Get start time from RunId
-      long programStartSecs = RunIds.getTime(RunIds.fromString(programRunId.getRun()), TimeUnit.SECONDS);
-      keysWithoutVersion.add(Fields.longField(StoreDefinition.AppMetadataStore.RUN_START_TIME,
-                                              getInvertedTsKeyPart(programStartSecs)));
-      keysWithoutVersion.add(Fields.stringField(StoreDefinition.AppMetadataStore.RUN_FIELD, programRunId.getRun()));
-      allKeys.add(keysWithoutVersion);
+      allKeys.add(getProgramRunInvertedTimeKey(TYPE_RUN_RECORD_COMPLETED, programRunId,
+                                               RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS)));
     }
 
-    return getRunsByKeys(allKeys);
+    return getRunsByFullKeys(allKeys);
   }
 
-  private Map<ProgramRunId, RunRecordDetail> getRunsByKeys(List<List<Field<?>>> allKeys) throws IOException {
-    Collection<Range> ranges = allKeys.stream().map(Range::singleton).collect(Collectors.toList());
-
-    try (CloseableIterator<StructuredRow> iterator =
-           getRunRecordsTable().multiScan(ranges, Integer.MAX_VALUE)) {
-      return StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, Spliterator.ORDERED), false)
-          .map(AppMetadataStore::deserializeRunRecordMeta)
-          .collect(Collectors.toMap(RunRecordDetail::getProgramRunId, r -> r, (r1, r2) -> {
-            throw new IllegalStateException("Duplicate run record for " + r1.getProgramRunId());
-          }, LinkedHashMap::new));
-    }
+  private Map<ProgramRunId, RunRecordDetail> getRunsByFullKeys(List<List<Field<?>>> keys) throws IOException {
+    return getRunRecordsTable().multiRead(keys).stream()
```

Comment:
@albertshau This is the performance fix. I tried to make the PR smaller... But I had to introduce the run reference Class and have a new interface `getRun(ProgramRunReference runRef)` in Store and apply it correctly in LogHandler and ProgramHander.

Reply:
I don't really get it. Is the issue that `getRun(ProgramRunId)` is slow? If so, why would adding a ProgramRunReference help vs just fixing the implementation of the existing method?

Reply:
So yes, the issue was `getRun(ProgramRunId)` being slow, it was using multiScan which inside the noSQL table scans ranges ignoring versions. (We made that change in LCM work because LogHandler does not pass in version, we could scan instead, runID is uuid). 

The fix is reverting the scan back to `mutiRead` by full keys, this way noSQL is back to efficient reads. For cases like LogHandler, we need a new interface to scan by run reference